Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/exposureriskmetadata/ExposureRiskMetadataDonor.kt

```diff
@@ -0,0 +1,75 @@
+package de.rki.coronawarnapp.datadonation.analytics.modules.exposureriskmetadata
+
+import de.rki.coronawarnapp.datadonation.analytics.AnalyticsSettings
+import de.rki.coronawarnapp.datadonation.analytics.modules.DonorModule
+import de.rki.coronawarnapp.risk.RiskLevelResult
+import de.rki.coronawarnapp.risk.RiskState
+import de.rki.coronawarnapp.risk.storage.RiskLevelStorage
+import de.rki.coronawarnapp.risk.tryLatestResultsWithDefaults
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaData
+import kotlinx.coroutines.flow.first
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class ExposureRiskMetadataDonor @Inject constructor(
+    private val riskLevelStorage: RiskLevelStorage,
+    private val analyticsSettings: AnalyticsSettings
+) : DonorModule {
+
+    override suspend fun beginDonation(request: DonorModule.Request): DonorModule.Contribution {
+        val previousMetadata = analyticsSettings.previousExposureRiskMetadata.value
+
+        val lastRiskResult = riskLevelStorage
+            .latestAndLastSuccessful
+            .first()
+            .tryLatestResultsWithDefaults()
+            .lastCalculated
+
+        val riskLevelForMetadata = lastRiskResult.calculateMetadataRiskLevel()
+
+        val newMetadata = PpaData.ExposureRiskMetadata.newBuilder()
+            .setRiskLevel(riskLevelForMetadata)
+            .setRiskLevelChangedComparedToPreviousSubmission(previousMetadata?.riskLevel != riskLevelForMetadata)
+            .setMostRecentDateAtRiskLevel(lastRiskResult.calculatedAt.millis)
+            .setDateChangedComparedToPreviousSubmission(
+                previousMetadata?.mostRecentDateAtRiskLevel != lastRiskResult.calculatedAt.millis
+            )
+            .build()
+
+        return ExposureRiskMetadataContribution(
+            contributionProto = newMetadata,
+            onContributionFinished = {
```

Comment:
I'm wondering if we need additional guards on `previousExposureRiskMetadata` to prevent threading issues.

Reply:
hmm, i'm not sure, technically analytics collections only happen every 24 hours right, so we should not be able to call this in succession often enough to be a problem right?